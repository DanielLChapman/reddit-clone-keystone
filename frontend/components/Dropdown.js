import React, { useState } from 'react';
import PropTypes from 'prop-types'

function Dropdown(props) {
    /*
        let options = ['Hot', 'Best', 'New', 'Top', 'Rising'];
        const [dropValue, setDropValue] = useState('Hot');
        <Dropdown returnFunc={setDropValue} selected={dropValue} options={options} />    
    */
    return (
        <select value={props.selected} onChange={(e) => {
            props.returnFunc(e.target.value);
        }}>
            {
                props.options && props.options.map((x, i) => (
                    <option key={i} value={x}>{x}</option>
                ))
            }
        </select>
    );
}

Dropdown.propTypes = {
    returnFunc: PropTypes.func.isRequired,
    options: PropTypes.array.isRequired,
    selected: PropTypes.string.isRequired,
};
  

export default Dropdown;