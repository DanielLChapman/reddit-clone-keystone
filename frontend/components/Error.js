import React from 'react';

function Error(props) {
    return (
        <span>
            {props.error}
        </span>
    );
}

export default Error;