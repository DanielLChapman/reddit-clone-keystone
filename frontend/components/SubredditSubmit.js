import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import React from 'react';
import useForm from '../lib/useForm';
import Router from 'next/router';
import Form from './styles/SubredditForm';


//NEED TO SET OWNER HERE
export const CREATE_SUBREDDIT_MUTATION = gql`
    mutation CREATE_SUBREDDIT_MUTATION(
        $name: String!,
        $slug: String!,
        $description: String,
        $sidebar: String,
        $title: String!
    ) {
        createSubreddit(
            data: {
                name: $name,
                slug: $slug,
                description: $description,
                sidebar: $sidebar,
                title: $title
            }
        ) {
            id
            name
        }
    }
`;


function SubredditSubmitForm(props) {
    const { inputs, handleChange, resetForm } = useForm({
        name: '',
        slug: '',
        description: '',
        sidebar: '',
        title: '',
        errors: {
            name: 'false'
        }
      });

      const [createSubreddit, {loading, data, error}] = useMutation(CREATE_SUBREDDIT_MUTATION);


    
    return (
        <div>
            <Form
            onSubmit={async (e) => {
                e.preventDefault();
                //form validation
                let format = /[ `!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?~]/;
                if (format.test(inputs.name)) {
                    alert('No spaces or special characters in name field');
                    let target = {
                        target: {
                            value:'true', 
                            name: 'name', 
                            type: 'error'
                        }
                    }
                    handleChange(target);
                    return;
                } 

                const res = await createSubreddit({
                    variables: {
                        name: inputs.name,
                        description: inputs.description,
                        sidebar: inputs.sidebar,
                        title: inputs.title,
                        slug: inputs.name.toLowerCase()
                    }
                })

                Router.push({
                    pathname: `/r/${inputs.name.toLowerCase()}`,
                  });

            }}>
                <fieldset>
                    <label htmlFor="name">
                        Name
                        <br />
                        <span className="lower-form-text">no spaces or special characters.</span>
                        <input 
                            required
                            type="text"
                            id="name"
                            name="name"
                            className={inputs.errors.name === 'true' ? 'error-field' : ''}
                            value={inputs.name}
                            placeholder="name"
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="title">
                        Title <br />
                        <input 
                            required
                            type="text"
                            id="title"
                            name="title"
                            placeholder="title"
                            value={inputs.title}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="description">
                        Description <br />
                        <textarea
                            id="description"
                            name="description"
                            placeholder="description"
                            value={inputs.description}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="sidebar">
                        Sidebar <br />
                        <textarea
                            id="sidebar"
                            name="sidebar"
                            placeholder="sidebar"
                            value={inputs.sidebar}
                            onChange={handleChange}
                        />
                    </label>
                    <button type="submit">Create Subreddit</button>
                    
                </fieldset>

            </Form>
        </div>
    );
}

export default SubredditSubmitForm;