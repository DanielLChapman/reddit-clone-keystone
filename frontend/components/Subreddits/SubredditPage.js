import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import React from 'react';
import SubredditContent from './SubredditContent';

const GET_SUBREDDIT_INFO = gql`
    query GET_SUBREDDIT_INFO($slug: String!) {
        allSubreddits(where: {
            slug: $slug
        }) {
            name
            title
            sidebar
            description
            status
            owner {
                name
            }
            moderators {
                name
            }
        }
    }
`;

function SubredditPage(props) {

    const {data, error, loading} = useQuery(GET_SUBREDDIT_INFO, {
        variables: {
            slug: props.slug
        }
    });
    if (loading) return <span>Loading...</span>
    if(error) return <span>Nothing To See Here</span>

    console.log(data);
    return (
        <div className="subreddit-content">
            <div className="subreddit-left">
                {/* HArd coding news id, will grab from gql later */}
                <SubredditContent slug={props.slug} type={props.type} selftext={props.selftext} id='60944706438ea508cdf752c9' /> 

            </div>
            <div className="subreddit-right">
                blah
            </div>
        </div>
    );
}

export default SubredditPage;