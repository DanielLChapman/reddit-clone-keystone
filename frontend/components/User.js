import { gql, useQuery } from '@apollo/client';

export const CURRENT_USER_QUERY = gql`
  query {
    authenticatedItem {
      ... on User {
        id
        username
        email
        name
        posts {
          votes {
            vflag
          }
        }
        comments {
          votes {
            vflag
          }
        }
        moderating {
          id
          slug
          name
        }
        owner {
          id
          slug
          name
        }
        subreddits {
          id
          slug
          name
        }
        postvotes {
          post {
            title
            id
          }
          id
          vflag
        }
        commentvotes {
          comment {
            id
          }
          id
          vflag
        }
    }
  }
  }
`;

export function useUser() {

  const { data } = useQuery(CURRENT_USER_QUERY);
  let test = {...data};

  
  if (test.authenticatedItem) {
    let x = Object.assign({}, test.authenticatedItem);
    //count up post votes
    let postVotes = 0;
    let commentVotes = 0;
    
    x.posts.forEach((q) => {
      q.votes.forEach((y) => {
        y.vflag === 'Upvote' ? postVotes += 1 : postVotes -= 1;
      })
    });
    //count up comment votes
    x.comments.forEach((q) => {
      q.votes.forEach((y) => {
        y.vflag === 'Upvote' ? commentVotes += 1 : commentVotes -= 1;
      })
    });

    x.postVotes = postVotes;
    x.commentVotes = commentVotes;

    return x;
  } else {
    return data?.authenticatedItem;
  }

  
}
