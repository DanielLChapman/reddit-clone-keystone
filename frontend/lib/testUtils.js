import casual from 'casual';

  
// seed it so we get consistent results
casual.seed(777);

const fakeUser = (overrides) => (
    {
    __typename: 'User',
    id: '4234',
    name: casual.name,
    username: casual.name,
    owner: [],
    email: casual.email,
    permissions: ['ADMIN'],
    moderating: [],
    subreddits: [],
    posts: [],
    postvotes: [],
    firstSignin: false,
    comments: [],
    commentvotes:[],
    ...overrides,
});

const fakeSubreddit = (overrides) => (
    {
        __typename: 'Subreddit',
        id: casual.uuid,
        name: casual.string,
        title: casual.string,
        slug: casual.string,
        sidebar: casual.string,
        rules: casual.string,
        description: casual.string,
        status: 'PUBLIC',
        ownder: fakeUser(),
        moderators: [],
        subscriber: [],
        posts: [],
        createdAt: Date.now(),
    }
);

const fakeLinkPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'www.google.com',
    type: 'Link',
    comments: [],
    removed: 'False',
    id: casual.uuid,


});

const fakeMediaPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: 'https://cdn.fileinfo.com/img/ss/sm/png_79.jpg',
    type: 'Link',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeTextPost = (overrides) => (
    {
    __typename: 'Post',
    title: casual.sentence,
    content: casual.sentence,
    user: fakeUser(),
    subreddit: fakeSubreddit(),
    createdAt: Date.now(),
    votes: [],
    post_slug: casual.string,
    link: '',
    type: 'Text',
    comments: [],
    id: casual.uuid,
    removed: 'False'

});

const fakeComment = (overrides) => (
    {
    __typename: 'Comment',
    content: casual.sentence,
    user: fakeUser(),
    post: fakeTextPost(),
    createdAt: Date.now(),
    votes: [],
    parent: null,
    id: casual.uuid,
    children: [],

});


// Fake LocalStorage
class LocalStorageMock {
    constructor() {
      this.store = {};
    }
  
    clear() {
      this.store = {};
    }
  
    getItem(key) {
      return this.store[key] || null;
    }
  
    setItem(key, value) {
      this.store[key] = value.toString();
    }
  
    removeItem(key) {
      delete this.store[key];
    }
  }



export {

    LocalStorageMock,
    fakeLinkPost,
    fakeComment,
    fakeMediaPost,
    fakeSubreddit,
    fakeTextPost,
    fakeUser,

  };