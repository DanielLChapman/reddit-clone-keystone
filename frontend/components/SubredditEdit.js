import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import React from 'react';
import useForm from '../lib/useForm';
import Router from 'next/router';
import Form from './styles/SubredditForm';
import { useState } from 'react';
import {GET_SUBREDDIT_INFO } from '../components/Subreddits/SubredditPage';
import Link from 'next/link';

//get subreddit info

//able to change owner needs to be added

export const EDIT_SUBREDDIT_MUTATION = gql`
    mutation EDIT_SUBREDDIT_MUTATION(
        $name: String!,
        $slug: String!,
        $description: String,
        $sidebar: String,
        $title: String!,
        $id: ID!
    ) {
        updateSubreddit(id: $id, data: {
            name: $name,
            slug: $slug,
            description: $description
            sidebar: $sidebar,
            title: $title
        }) {
            id
        }
    }
`;

function SubredditEdit(props) {
    const [isChecked, setIsChecked] = useState(props.subreddit.status === 'PUBLIC' ? true : false);

    const { inputs, handleChange, resetForm } = useForm({
        name: props.subreddit.name,//shouldn't be able to be changed so set as data from query
        slug: props.subreddit.slug,//shouldn't be able to be changed so set as data from query
        description: props.subreddit.description,
        sidebar: props.subreddit.sidebar,
        title: props.subreddit.title,
        errors: {
            name: 'false'
        }
      });

      const [updateSubreddit, {loading, data, error}] = useMutation(EDIT_SUBREDDIT_MUTATION);

    
    if (error) return <div>Error</div>;
    if (loading) return <span>Loading...</span>
    return (
        <div>
            {
                data && (
                    <section className="success-form-message">
                        Successfully updated. <a href={`/r/${props.subreddit.slug}`}>View now </a>
                    </section>
                )
            }
            
            <Form 
            onSubmit={async (e) => {
                e.preventDefault();
                //form validation

                const res = await updateSubreddit({
                    variables: {
                        id: props.subreddit.id,
                        name: inputs.name,
                        description: inputs.description,
                        sidebar: inputs.sidebar,
                        title: inputs.title,
                        slug: inputs.name.toLowerCase()
                    }
                })


            }}>
                <fieldset>
                
                    <label htmlFor="title">
                        Title <br />
                        <input 
                            required
                            type="text"
                            id="title"
                            name="title"
                            value={inputs.title}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="description">
                        Description <br />
                        <textarea
                            id="description"
                            name="description"
                            placeholder="description"
                            value={inputs.description}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="sidebar">
                        Sidebar <br />
                        <textarea
                            id="sidebar"
                            name="sidebar"
                            placeholder="sidebar"
                            value={inputs.sidebar}
                            onChange={handleChange}
                        />
                    </label>
                    <label htmlFor="status" className="form-checkbox">
                        Status: {isChecked ? 'Public' : 'Private'}<br />
                        <input
                            name="status"
                            type="checkbox"
                            checked={isChecked}
                            onChange={() => {setIsChecked(!isChecked) }} />
                    </label>
                    <button type="submit">Edit {/* subreddit name */}</button>
                    
                </fieldset>

            </Form>
        </div>
    );
}

export default SubredditEdit;