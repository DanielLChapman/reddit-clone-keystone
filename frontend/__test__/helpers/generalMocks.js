import { CURRENT_USER_QUERY } from '../../components/User';
import { fakeTextPost, fakeUser} from '../../lib/testUtils';

const fakeUserObj = fakeUser();
const notSignedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem: null } },
    },
  ];
  
const signedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem: fakeUser() /*{
        
            id: fakeUserObj.id,
            username: fakeUserObj.username,
            email: fakeUserObj.email,
            name:fakeUserObj.name,
            posts: [],
            comments: [],
            moderating: [],
            owner:[],
            subreddits:[],
            postvotes:[],
            commentvotes: [],
        
      }*/ } },
    },
  ];
  
const signedInMocksWithPosts = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: {
        data: {
          authenticatedItem: fakeUser({
            posts: [fakeTextPost()],
          }),
        },
      },
    },
];

export {
    notSignedInMocks,
    signedInMocks,
    signedInMocksWithPosts
}

/* 

{
        id
        username
        email
        name
        posts {
          votes {
            vflag
          }
        }
        comments {
          votes {
            vflag
          }
        }
        moderating {
          id
          slug
          name
        }
        owner {
          id
          slug
          name
        }
        subreddits {
          id
          slug
          name
        }
        postvotes {
          post {
            title
            id
          }
          id
          vflag
        }
        commentvotes {
          comment {
            id
          }
          id
          vflag
        }
    }

    */