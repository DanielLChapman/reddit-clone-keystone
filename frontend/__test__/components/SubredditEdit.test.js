import { act, render, screen, waitFor } from '@testing-library/react';
import {MockedProvider} from '@apollo/react-testing';
import { signedInMocks, signedInMocksWithPosts } from '../helpers/generalMocks';
import wait from 'waait';
import { fakeSubreddit, fakeUser } from '../../lib/testUtils';
import userEvent from '@testing-library/user-event';
import { CREATE_SUBREDDIT_MUTATION } from '../../components/SubredditSubmit';
import  Router  from 'next/router';
import SubredditEditPage from '../../pages/r/[slug]/edit';
import { GET_SUBREDDIT_INFO } from '../../components/Subreddits/SubredditPage';
import { CURRENT_USER_QUERY } from '../../components/User';
import { EDIT_SUBREDDIT_MUTATION } from '../../components/SubredditEdit';

const subreddit = fakeSubreddit();
const user = fakeUser();

jest.mock('next/router', () => ({
    push: jest.fn(),
}));

const mocks = [
    // Mutation Mock
    {
      request: {
        query: GET_SUBREDDIT_INFO,
        variables: {
          slug: subreddit.name.toLowerCase(),
        },
      },
      result: () => {
        return {
      data: { allSubreddits: [{ 

          id: subreddit.id,
          name: subreddit.name,
          title: subreddit.title,
          slug: subreddit.name.toLowerCase(),
          sidebar: subreddit.sidebar,
          description: subreddit.description,
          status: 'Public',
          owner: {
              username:subreddit.owner.username,
          },
          moderators: [],
          createdAt: subreddit.createdAt,
      }] },
    }},
    },

  ];
  
  const signedInCustomMocks = [
    ...signedInMocks,
    ...mocks
  ]

  let correctMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem:  {
          ...user,
          owner: [{
            id: subreddit.id,
            name: subreddit.name,
            slug: subreddit.name.toLowerCase()
          }]
      } } },
    },
    {
      request: {
        query: GET_SUBREDDIT_INFO,
        variables: {
          slug: subreddit.name.toLowerCase(),
        },
      },
      result: () => {
        return {
          data: { 
            allSubreddits: [{ 
              id: subreddit.id,
              name: subreddit.name,
              title: subreddit.title,
              slug: subreddit.name.toLowerCase(),
              sidebar: subreddit.sidebar,
              description: subreddit.description,
              status: subreddit.status,
              owner: {
                  username:subreddit.owner.username
              },
              moderators: [],
              createdAt: subreddit.createdAt,
            }] 
          },
        }
      },
    },
    {
      request: { query: EDIT_SUBREDDIT_MUTATION,
        variables: {
          id:subreddit.id,
          name:subreddit.name,
          description:subreddit.description,
          sidebar:subreddit.sidebar,
          title:subreddit.title+subreddit.title,
          slug: subreddit.name.toLowerCase(),
        } },
      result: { data: { updateSubreddit:  {
          id: subreddit.id,
          __typename: 'Subreddit'
      } } },
    },
]

/*
const EDIT_SUBREDDIT_MUTATION = gql`
    mutation EDIT_SUBREDDIT_MUTATION(
        $name: String!,
        $slug: String!,
        $description: String,
        $sidebar: String,
        $title: String!,
        $id: ID!
    ) {
        updateSubreddit(id: $id, data: {
            name: $name,
            slug: $slug,
            description: $description
            sidebar: $sidebar,
            title: $title
        }) {
            id
        }
    }
`;
*/

describe('it renders a subreddit submit page and submits successfully', () => {
    it ('does not load unless signed in', async () => {
        const { container, debug } = render(
            <MockedProvider mocks={mocks}>
              <SubredditEditPage query={{slug: subreddit.name.toLowerCase()}}/>
            </MockedProvider>
        );

        await screen.findByText(
            'Sign Into Your Account'
        );
        expect(container).toMatchSnapshot();
    });

    it ('loads when signed in, but not with the correct permissions', async () => {
        const { container, debug } = render(
            <MockedProvider mocks={signedInCustomMocks}>
              <SubredditEditPage query={{slug: subreddit.name.toLowerCase()}}/>
            </MockedProvider>
        );

        await screen.findByText(
            'You dont have the required permission for this.'
        );

        expect(container).toMatchSnapshot();
    });

    it ('loads when signed in', async () => {
      const { container, debug } = render(
          <MockedProvider mocks={correctMocks}>
            <SubredditEditPage query={{slug: subreddit.name.toLowerCase()}}/>
          </MockedProvider>
      );

      await screen.findByText('Title');

      expect(container).toMatchSnapshot();
  });

    it ('calls the mutation', async () => {
        const { container, debug } = render(
            <MockedProvider mocks={correctMocks}>
              <SubredditEditPage query={{slug: subreddit.name.toLowerCase()}} />
            </MockedProvider>
        );

        await screen.findByText('Title');
        await userEvent.type(screen.getByLabelText('Title'), subreddit.title);
        
        await userEvent.click(screen.getByText('Edit'));

        await waitFor(() => wait(0));

        expect(container).toHaveTextContent('Successfully updated.');
        
    });
})