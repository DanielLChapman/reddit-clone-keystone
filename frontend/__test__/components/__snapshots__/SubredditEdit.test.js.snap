// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a subreddit submit page and submits successfully does not load unless signed in 1`] = `
<div>
  <div>
    <form
      class="Form-sc-1tx35ym-0 etUUQp"
      method="POST"
    >
      <h2>
        Sign Into Your Account
      </h2>
      <fieldset>
        <label
          for="email"
        >
          Email
          <input
            autocomplete="email"
            name="email"
            placeholder="Your Email Address"
            type="email"
            value=""
          />
        </label>
        <label
          for="password"
        >
          Password
          <input
            autocomplete="password"
            name="password"
            placeholder="Password"
            type="password"
            value=""
          />
        </label>
        <button
          type="submit"
        >
          Sign In
        </button>
      </fieldset>
    </form>
    <style>
                      body, html{                background-color:white;                }            
    </style>
  </div>
</div>
`;

exports[`it renders a subreddit submit page and submits successfully loads when signed in 1`] = `
<div>
  <div>
    <div>
      <form
        class="SubredditForm__Form-q7oa8i-0 kqHjhW"
      >
        <fieldset>
          <label
            for="title"
          >
            Title 
            <br />
            <input
              id="title"
              name="title"
              required=""
              type="text"
              value="sit saepe assumenda deserunt consequatur itaque et"
            />
          </label>
          <label
            for="description"
          >
            Description 
            <br />
            <textarea
              id="description"
              name="description"
              placeholder="description"
            >
              quidem voluptatibus libero aut autem ut ullam
            </textarea>
          </label>
          <label
            for="sidebar"
          >
            Sidebar 
            <br />
            <textarea
              id="sidebar"
              name="sidebar"
              placeholder="sidebar"
            >
              labore qui aut inventore aut eveniet beatae
            </textarea>
          </label>
          <label
            class="form-checkbox"
            for="status"
          >
            Status: 
            Public
            <br />
            <input
              checked=""
              name="status"
              type="checkbox"
            />
          </label>
          <button
            type="submit"
          >
            Edit 
          </button>
        </fieldset>
      </form>
    </div>
    <style>
                      body, html{                background-color:white;                }            
    </style>
  </div>
</div>
`;

exports[`it renders a subreddit submit page and submits successfully loads when signed in, but not with the correct permissions 1`] = `
<div>
  <div>
    <div>
      <span>
        You dont have the required permission for this.
      </span>
    </div>
    <style>
                      body, html{                background-color:white;                }            
    </style>
  </div>
</div>
`;
